promql/parser/prettier.go

```diff
@@ -0,0 +1,562 @@
+	// is property is applied only after encountering a Left_Brace. This is dependent across
+	// multiple padder states and hence needs to be carried along.
+	labelsSplittable bool
+	isFirstLabel     bool
+}
+
+// Prettify prettifies the expression passed as an argument. It standardizes
+// the input expression, rearranges the lexical items and calls prettify.
+// The returned string is a formatted expression.
+func Prettify(expression string) (string, error) {
+	ptr := &prettier{
+		pd: &padder{indent: defaultIndent, isNewLineApplied: true, columnLimit: columnLimit},
+	}
+	standardizeExprStr := stringifyItems(
+		ptr.rearrangeItems(ptr.lexItems(expression)),
+	)
+	if err := ptr.parseExpr(standardizeExprStr); err != nil {
+		return expression, err
+	}
+	formattedExpr, err := ptr.prettify(ptr.lexItems(standardizeExprStr))
+	if err != nil {
+		return expression, errors.Wrap(err, "Prettify")
+	}
+	return formattedExpr, nil
+}
+
+// rearrangeItems rearranges the lexical items slice to ensure proper order
+// of items before prettifying the input expression.
+func (p *prettier) rearrangeItems(items []Item) []Item {
+	for i := 0; i < len(items); i++ {
+		item := items[i]
+		switch item.Typ {
+		case SUM, AVG, MIN, MAX, COUNT, COUNT_VALUES,
+			STDDEV, STDVAR, TOPK, BOTTOMK, QUANTILE:
+			// Case: sum(metric_name) by(job) => sum by(job) (metric_name)
+			var (
+				formatItems       bool
+				aggregatorIndex   = i
+				keywordIndex      = itemNotFound
+				closingParenIndex = itemNotFound
+				openBracketsCount = 0
+			)
+			for index := i; index < len(items); index++ {
+				it := items[index]
+				switch it.Typ {
+				case LEFT_PAREN, LEFT_BRACE:
+					openBracketsCount++
+				case RIGHT_PAREN, RIGHT_BRACE:
+					openBracketsCount--
+				}
+				if openBracketsCount < 0 {
+					break
+				}
+				if openBracketsCount == 0 && it.Typ.IsKeyword() {
+					keywordIndex = index
+					break
+				}
+			}
+			if keywordIndex == itemNotFound {
+				continue
+			}
+
+			if items[keywordIndex-1].Typ != COMMENT && keywordIndex-1 != aggregatorIndex {
+				formatItems = true
+			} else if items[keywordIndex-1].Typ == COMMENT {
+				// Peek back until no comment is found.
+				// If the item immediately before the comment is not aggregator, that means
+				// the keyword is not (immediately) after aggregator item and hence formatting
+				// is required.
+				var j int
+				for j = keywordIndex; j > aggregatorIndex; j-- {
+					if items[j].Typ == COMMENT {
+						continue
+					}
+					break
+				}
+				if j != aggregatorIndex {
+					formatItems = true
+				}
+			}
+			if formatItems {
+				var (
+					tempItems              []Item
+					finishedKeywordWriting bool
+				)
+				// Get index of the closing paren.
+				for j := keywordIndex; j <= len(items); j++ {
+					if items[j].Typ == RIGHT_PAREN {
+						closingParenIndex = j
+						break
+					}
+				}
+				if closingParenIndex == itemNotFound {
+					panic("invalid paren index: closing paren not found")
```

Comment:
Let's not panic anywhere here and rather return errors. The user of `Prettify` would not be expecting a panic :)